### ISSUE TITLE
How to visualize Gaussian primitives?

### ISSUE THREAD

**Issue by chenhaomingbob** [JuIvyy/GPOcc#7]
Thank you for the excellent paper and for open-sourcing this great work! The results are impressive and the codebase is well-organized. I would like to visualize the Gaussian primitives used in the model (as shown in Fig. 3). However, I could not find clear instructions or scripts in the repository for this purpose.

## Questions
1. Is there an existing script or function to visualize the Gaussian primitives?
2. What is the recommended way to inspect or plot these representations for debugging or analysis?
3. Are there any specific dependencies or visualization tools (e.g., matplotlib, open3d, wandb) that should be used?

Thank you again for your great work! Looking forward to your guidance.

**Comment by JuIvyy** [comment]
Hi, you can refer to the code below to implement Gaussian visualization.

This function visualizes each Gaussian as an ellipsoid using its predicted mean, scale, rotation, opacity, and semantic label. It also supports drawing the camera frustum and optionally matching the Matplotlib view to the default Plotly camera view.

```python
def draw_gaussian(
    vox_near, scene_size, save_path, gaussian, idx=0, sem=True, K_idx=0, scene_name=None,
    cam_pose=None, intrinsic=None, d=0.5, is_cam2world=True,
    plotly_like_when_none=True,
    # ===== 新增：让视角与 Plotly 一致 =====
    match_plotly_view=True,
    plotly_eye=(1.25, 1.25, 1.25),     # 对应 Plotly 默认 camera.eye
    plotly_orthographic=False          # 对应 draw_voxel_plotly_image(orthographic)
):
    import os
    import numpy as np
    import matplotlib.pyplot as plt
    from matplotlib import cm, colors
    from mpl_toolkits.mplot3d import Axes3D  # noqa: F401
    from pyquaternion import Quaternion
    import torch

    def _as_numpy(x):
        if isinstance(x, torch.Tensor):
            return x.detach().cpu().numpy()
        return np.asarray(x)

    # —— 新：把 Plotly 的 eye -> Matplotlib 的 elev/azim，并切换正交/透视 ——
    def _set_plotly_like_view(ax, eye=(1.25, 1.25, 1.25), orthographic=False):
        ex, ey, ez = float(eye[0]), float(eye[1]), float(eye[2])
        r = (ex**2 + ey**2 + ez**2) ** 0.5 + 1e-9
        elev = np.degrees(np.arcsin(ez / r))          # [-90, 90]
        azim = np.degrees(np.arctan2(ey, ex))         # (-180, 180]
        ax.view_init(elev=elev, azim=azim)
        # Matplotlib 3.2+ 支持正交投影
        if hasattr(ax, "set_proj_type"):
            ax.set_proj_type('ortho' if orthographic else 'persp')

    def _draw_camera(ax, cam_pose, intrinsic=None, d=0.5, is_cam2world=True, color='k'):
        C = _as_numpy(cam_pose)
        if is_cam2world:
            R = C[:3, :3]; t = C[:3, 3]
        else:
            Ci = np.linalg.inv(C); R = Ci[:3, :3]; t = Ci[:3, 3]

        if intrinsic is not None:
            K = _as_numpy(intrinsic)
            fx, fy, cx, cy = K[0, 0], K[1, 1], K[0, 2], K[1, 2]
            y = d * (2 * cy) / (2 * fy)
            x = d * (2 * cx) / (2 * fx)
            cam_pts = np.array([[0, 0, 0],
                                [ x,  y, d], [-x,  y, d],
                                [-x, -y, d], [ x, -y, d]], dtype=float)
            cam_pts_w = (R @ cam_pts.T).T + t[None, :]
            for i, j in [(0,1),(0,2),(0,3),(0,4),(1,2),(2,3),(3,4),(4,1)]:
                ax.plot([cam_pts_w[i,0], cam_pts_w[j,0]],
                        [cam_pts_w[i,1], cam_pts_w[j,1]],
                        [cam_pts_w[i,2], cam_pts_w[j,2]], color=color, linewidth=1.)
        else:
            scale = 0.2 * float(np.linalg.norm(scene_size))
            origin = t
            axes = np.stack([R @ np.array([1,0,0]),
                             R @ np.array([0,1,0]),
                             R @ np.array([0,0,1])], 0)
            cols = ['r','g','b']
            for a, c in zip(axes, cols):
                p = origin + a * scale
                ax.plot([origin[0], p[0]], [origin[1], p[1]], [origin[2], p[2]], c=c, linewidth=2)
            ax.scatter(origin[0], origin[1], origin[2], c='k', s=8)

    # --- 原有数据处理 ---
    empty_label = 12
    sem_cmap = np.array([
        [0, 0, 0],
        [214, 38, 40], [43, 160, 4], [158, 216, 229], [114, 158, 206],
        [204, 204, 91], [255, 186, 119], [147, 102, 188], [30, 119, 181],
        [160, 188, 33], [255, 127, 12], [196, 175, 214], [128, 128, 128],
    ]) / 255.0

    means = gaussian['means'][idx].detach().cpu().numpy()
    scales = gaussian['scales'][idx].detach().cpu().numpy()
    rotations = gaussian['rotations'][idx].detach().cpu().numpy()
    opas = gaussian['opacities'][idx]
    if opas.numel() == 0:
        opas = torch.ones_like(gaussian.means[idx][..., :1])
    opas = opas.squeeze().detach().cpu().numpy()
    sems = gaussian['semantics'][idx].detach().cpu().numpy()
    pred = np.argmax(sems, axis=-1)

    mask = (pred != empty_label)
    means, scales, rotations, opas, pred = means[mask], scales[mask], rotations[mask], opas[mask], pred[mask]

    ellipNumber = means.shape[0]
    norm = colors.Normalize(vmin=-1.0, vmax=5.4)
    cmap = cm.jet
    m = cm.ScalarMappable(norm=norm, cmap=cmap)

    fig = plt.figure(figsize=(9, 9), dpi=300)
    ax = fig.add_subplot(111, projection='3d')

    vox_near = _as_numpy(vox_near).reshape(3)
    scene_size = _as_numpy(scene_size).reshape(3)

    scalar = 0.8

    # ---------- 计算绘制对象 +（可选）相机视锥的联合包围盒 ----------
    if means.size:
        radii = scalar * scales
        gauss_min = (means - radii).min(axis=0)
        gauss_max = (means + radii).max(axis=0)
    else:
        gauss_min, gauss_max = vox_near, vox_near + scene_size

    if cam_pose is not None and intrinsic is not None:
        K = _as_numpy(intrinsic)
        fx, fy, cx, cy = K[0, 0], K[1, 1], K[0, 2], K[1, 2]
        y = d * (2 * cy) / (2 * fy)
        x = d * (2 * cx) / (2 * fx)
        cam_pts = np.array([[0, 0, 0], [ x,  y, d], [-x,  y, d], [-x, -y, d], [ x, -y, d]])
        C = _as_numpy(cam_pose)
        if is_cam2world:
            R = C[:3, :3]; t = C[:3, 3]
        else:
            Ci = np.linalg.inv(C); R = Ci[:3, :3]; t = Ci[:3, 3]
        cam_pts_w = (R @ cam_pts.T).T + t[None, :]
        all_min = np.minimum(gauss_min, cam_pts_w.min(axis=0))
        all_max = np.maximum(gauss_max, cam_pts_w.max(axis=0))
    else:
        all_min, all_max = gauss_min, gauss_max

    # —— 设定坐标范围（与 Plotly “fit 所有 trace”一致）——
    ax.set_xlim(all_min[0], all_max[0])
    ax.set_ylim(all_min[1], all_max[1])
    ax.set_zlim(all_min[2], all_max[2])

    # 等比例（和 Plotly aspectmode='data' 一致）
    sx, sy, sz = (all_max - all_min).tolist()
    ax.set_box_aspect((sx, sy, sz))

    # —— 设置 Matplotlib 视角 = Plotly 的 eye / orthographic —— 
    if match_plotly_view:
        _set_plotly_like_view(ax, eye=plotly_eye, orthographic=plotly_orthographic)
    else:
        ax.view_init(elev=30, azim=45)  # 保留你原来的默认

    # 椭球参数网格
    u = np.linspace(0.0, 2.0 * np.pi, 10)
    v = np.linspace(0.0, np.pi, 10)
    cu, su = np.cos(u), np.sin(u)
    sv, cv = np.sin(v), np.cos(v)
    X0 = np.outer(cu, sv)
    Y0 = np.outer(su, sv)
    Z0 = np.outer(np.ones_like(u), cv)

    for i in range(ellipNumber):
        center = means[i]
        radii = scales[i] * scalar
        R = Quaternion(rotations[i]).rotation_matrix.T
        xyz = np.stack([radii[0]*X0, radii[1]*Y0, radii[2]*Z0], axis=-1)
        xyz = (R[None, None, ...] @ xyz[..., None]).squeeze(-1) + center
        color = sem_cmap[pred[i]] if sem else m.to_rgba(center[2])
        ax.plot_surface(
            xyz[..., 0], xyz[..., 1], xyz[..., 2],
            rstride=1, cstride=1, color=color, linewidth=0,
            alpha=float(np.clip(opas[i], 0.0, 1.0)), shade=True
        )

    if cam_pose is not None:
        _draw_camera(ax, cam_pose, intrinsic=intrinsic, d=d, is_cam2world=is_cam2world, color='k')

    ax.set_axis_off()
    plt.savefig(save_path, bbox_inches='tight', pad_inches=0)
    plt.cla(); plt.clf()
```

Sorry that this snippet has not been fully cleaned up yet. You may need adapt to your code (use AI to help you).

**Comment by chenhaomingbob** [comment]
Thank you very much !!!
